File: code/dronecode/controller/navigation.cpp
#include "common.h"
#include "navigation.h"
#include "storage.h"
#include "motion.h"
#include "bluetooth.h"

#define ISGN(X) ((X)>=0?1:-1)
#define IABS(X) ((X)>=0?(X):(-(X)))
#define IMIN(X, Y) ((X)>(Y)?(Y):(X))
#define RANGE_LIMIT(INP, LIM) IMIN((LIM), IABS(INP))* ISGN(INP)
#define THRUST_NAVIGATION_THRESHOLD 900

struct PidState
{
    float lasterror;
    float i;
    float target;

    void reset(float target_)
    {
        lasterror = 0;
        i = 0;
        this->target = target_;
    }
};

PidState yaw_pid, roll_pid, pitch_pid;

void reset_pids(float yaw = 0, float pitch = 0, float roll = 0)
{
    yaw_pid.reset(yaw);
    pitch_pid.reset(0);
    roll_pid.reset(0);
}

void init_navigation()
{
    thrust_input = 0;
    yaw_input = 0;
    pitch_input = 0;
    roll_input = 0;
    reset_pids();
}

void pid(const float e0,
         const float e1,
         const float dt,
         const float kp,
         const float ki,
         const float kd,
         const float limit,
         float &i,
         float &o)
{
    i += e1 * dt * ki;
    i = RANGE_LIMIT(i, limit);


    float d = (e1 - e0) / dt * kd;

    float p = kp * e1;
    o = p + i + d;
    o = RANGE_LIMIT(o, limit);
}

float fit_angle(float angle)
{
    if (angle > 180)
    {
        return -(360 - angle);
    }
    return angle;
}

float angle_error(float target, float value)
{
    return target - value;
}

uint32_t prev_t = 0;
int x;

void push_float(float q)
{
    x = (x + 1) & 7;
    if (x == 0)
    {
        bt_send_float(1, q);
    }
}

uint16_t sensordump_ptr = 0;
float sensordump[1000];
bool flush_sensor_dump_next = false;

void flush_sensordump()
{
    if (!motors_enabled)
    {
        flush_sensor_dump_next = true;
    }
}

void do_flush_sensordump()
{
    if (!motors_enabled)
    {
        if (flush_sensor_dump_next)
        {
            for (uint16_t i = 0; i < 1000; i++)
            {
                bt_send_float(1, sensordump[i]);
                wdt_reset();
                delay(2);
            }
        }
        flush_sensor_dump_next = false;
    }
    else
    {
        flush_sensor_dump_next = false;
    }
}

void start_sensorlog()
{
    sensordump_ptr = 0;
    sensordump[999] = 0;
}

void push_point(float q, float dt)
{
    if (sensordump_ptr < 999)
    {
        sensordump[sensordump_ptr] = q;
        sensordump_ptr += 1;
        sensordump[999] += dt / 999;
    }
}

void navigate()
{
    constexpr float pid_scale_factor = 0.01;
    constexpr float input_scale_factor = 0.1;
    const int16_t command_limit = thrust_input >> 2;
    float yaw_measured, pitch_measured, roll_measured;
    get_ypr(yaw_measured, pitch_measured, roll_measured);
    uint32_t t = micros();
    uint32_t delta_t;
    if (t > prev_t)
    {
        delta_t = t - prev_t;
    }
    else
    {
        delta_t = (4294967295 - prev_t) + t;
    }
    prev_t = t;
    float dt = delta_t * 0.000001f;
    if (motors_enabled)
    {
        push_point(pitch_measured, dt);
    }
    else
    {
        do_flush_sensordump();
    }


    motor_thrust = thrust_input;
    motor_pitch = 0;
    motor_roll = 0;
    motor_yaw = 0;
    //if(motors_enabled) push_float(pitch_measured);
    if (thrust_input < THRUST_NAVIGATION_THRESHOLD)
    {
        reset_pids(0, 0, 0);
        motor_yaw = RANGE_LIMIT(0, command_limit);
        motor_pitch = RANGE_LIMIT(0, command_limit);
        motor_roll = RANGE_LIMIT(0, command_limit);
    }
    else
    {
        float yaw_error = yaw_pid.target + yaw_input - yaw_measured;
        float yaw_pid_result;
        pid(yaw_pid.lasterror,
            yaw_error,
            dt,
            yaw_kp * pid_scale_factor,
            yaw_ki * pid_scale_factor,
            yaw_kd * pid_scale_factor,
            command_limit,
            yaw_pid.i,
            yaw_pid_result);
        yaw_pid.lasterror = yaw_error;

        float pitch_error = pitch_pid.target - pitch_measured + pitch_input;
        float pitch_pid_result;
        pid(pitch_pid.lasterror,
            pitch_error,
            dt,
            pitch_kp * pid_scale_factor,
            pitch_ki * pid_scale_factor,
            pitch_kd * pid_scale_factor,
            command_limit,
            pitch_pid.i,
            pitch_pid_result);
        pitch_pid.lasterror = pitch_error;

        float roll_error = roll_pid.target + -roll_measured + roll_input;
        float roll_pid_result;
        pid(roll_pid.lasterror,
            roll_error,
            dt,
            roll_kp * pid_scale_factor,
            roll_ki * pid_scale_factor,
            roll_kd * pid_scale_factor,
            command_limit,
            roll_pid.i,
            roll_pid_result);
        roll_pid.lasterror = roll_error;

        motor_yaw = RANGE_LIMIT(yaw_pid_result, command_limit);
        motor_pitch = RANGE_LIMIT(pitch_pid_result, command_limit);
        motor_roll = RANGE_LIMIT(roll_pid_result, command_limit);
    }
}

File: code/src/navigation.rs
use embassy_time::Instant;

use crate::{crsf::CRSFChannels, icm42688::ICM42688, motors::MotorInputs};

struct Pid {
    i: f32,
    last_error: f32,
}

impl Pid {
    fn new() -> Self {
        Pid {
            i: 0f32,
            last_error: 0f32,
        }
    }

    fn calculate(&mut self, kp: f32, ki: f32, kd: f32, error: f32, dt: f32, i_limit: f32) -> f32 {
        self.i += error * dt * ki;

        self.i = if self.i > i_limit {
            i_limit
        } else if self.i < -i_limit {
            -i_limit
        } else {
            self.i
        };

        let result = kp * error + self.i + (error - self.last_error) * kd / dt;
        self.last_error = error;

        return result;
    }

    fn reset(&mut self) {
        self.last_error = 0f32;
        self.i = 0f32;
    }
}

pub struct NavigationContext {
    yaw_pid: Pid,
    pitch_pid: Pid,
    roll_pid: Pid,

    last_pid_time: Instant,
}

impl NavigationContext {
    pub fn new() -> Self {
        NavigationContext {
            yaw_pid: Pid::new(),
            pitch_pid: Pid::new(),
            roll_pid: Pid::new(),
            last_pid_time: Instant::now(),
        }
    }

    fn reset(&mut self) {
        self.yaw_pid.reset();
        self.pitch_pid.reset();
        self.roll_pid.reset();
        self.last_pid_time = Instant::now();
    }
}

fn range_limit(value: f32, limit: f32) -> f32 {
    if value > limit {
        limit
    } else if value < -limit {
        -limit
    } else {
        value
    }
}

pub fn navigate(
    imu: &mut ICM42688,
    context: &mut NavigationContext,
    inputs: &CRSFChannels,
) -> MotorInputs {
    const YAW_OFFSET: f32 = 0.47141057;
    const PITCH_OFFSET: f32 = 0.3117653;
    const ROLL_OFFSET: f32 = 0.088096835;
    let (yaw_measured, pitch_measured, roll_measured) = imu.get_ypr_deg();

    let yaw_measured = yaw_measured - YAW_OFFSET;
    let pitch_measured = pitch_measured - PITCH_OFFSET;
    let roll_measured = roll_measured - ROLL_OFFSET;

    let roll_measured = roll_measured * -1f32;

    let yaw_error = (inputs.yaw() as f32) - yaw_measured;
    let pitch_error = (inputs.pitch() as f32) - pitch_measured;
    let roll_error = (inputs.roll() as f32) - roll_measured;

    let thrust = inputs.throttle() as f32;
    let command_limit = thrust / 4f32;

    let motor_thrust = inputs.throttle();

    let now = Instant::now();
    let dt = (now - context.last_pid_time).as_micros() as f32 * 1e-6;
    context.last_pid_time = now;

    let p_scale_factor = 1.0f32 + (inputs.aux1() / 2) as f32 / 8f32;

    let d = (inputs.aux2() / 2) as f32 / 2.0f32;

    if motor_thrust < 200 {
        context.reset();

        return MotorInputs::idle(motor_thrust);
    } else {
        let yaw_input =
            context
                .yaw_pid
                .calculate(100f32, 72f32, 0f32, yaw_error, dt, command_limit);

        let yaw_input = range_limit(yaw_input, command_limit) as i16;

        let pitch_input = context.pitch_pid.calculate(
            12.6f32 * p_scale_factor,
            60f32,
            d,
            pitch_error,
            dt,
            command_limit,
        );
        let pitch_input = range_limit(pitch_input, command_limit) as i16;

        let roll_input = context.roll_pid.calculate(
            12.6f32 * p_scale_factor,
            60f32,
            d,
            roll_error,
            dt,
            command_limit,
        );
        let roll_input = range_limit(roll_input, command_limit) as i16;

        return MotorInputs {
            motor_thrust: motor_thrust,
            yaw_input: yaw_input,
            pitch_input: pitch_input,
            roll_input: roll_input,
        };
    }
}
